packages/common/infra/src/workspace/manager.ts

```diff
@@ -0,0 +1,190 @@
+   * need the workspace anymore.
+   */
+  open(metadata: WorkspaceMetadata): {
+    workspace: Workspace;
+    release: () => void;
+  } {
+    const exist = this.pool.get(metadata.id);
+    if (exist) {
+      return {
+        workspace: exist.obj,
+        release: exist.release,
+      };
+    }
+
+    const workspace = this.instantiate(metadata);
+    const ref = this.pool.put(workspace.meta.id, workspace);
+
+    return {
+      workspace: ref.obj,
+      release: ref.release,
+    };
+  }
+
+  createWorkspace(
+    flavour: WorkspaceFlavour,
+    initial?: (
+      workspace: BlockSuiteWorkspace,
+      blobStorage: BlobStorage
+    ) => Promise<void>
+  ): Promise<WorkspaceMetadata> {
+    logger.info(`create workspace [${flavour}]`);
+    return this.list.create(flavour, initial);
+  }
+
+  /**
+   * delete workspace by metadata, same as `WorkspaceList.deleteWorkspace`
+   */
+  async deleteWorkspace(metadata: WorkspaceMetadata) {
+    await this.list.delete(metadata);
+  }
+
+  /**
+   * helper function to transform local workspace to cloud workspace
+   */
+  async transformLocalToCloud(local: Workspace): Promise<WorkspaceMetadata> {
+    assertEquals(local.flavour, WorkspaceFlavour.LOCAL);
+
+    await local.engine.sync.waitForSynced();
+
+    const newId = await this.list.create(
+      WorkspaceFlavour.AFFINE_CLOUD,
+      async (ws, bs) => {
+        applyUpdate(ws.doc, encodeStateAsUpdate(local.blockSuiteWorkspace.doc));
+
+        for (const subdoc of local.blockSuiteWorkspace.doc.getSubdocs()) {
+          for (const newSubdoc of ws.doc.getSubdocs()) {
+            if (newSubdoc.guid === subdoc.guid) {
+              applyUpdate(newSubdoc, encodeStateAsUpdate(subdoc));
+            }
+          }
+        }
+
+        const blobList = await local.engine.blob.list();
+
+        for (const blobKey of blobList) {
+          const blob = await local.engine.blob.get(blobKey);
+          if (blob) {
+            await bs.set(blobKey, blob);
+          }
+        }
+      }
+    );
+
+    await this.list.delete(local.meta);
+
+    return newId;
+  }
+
+  /**
+   * helper function to get blob without open workspace, its be used for download workspace avatars.
+   */
+  getWorkspaceBlob(metadata: WorkspaceMetadata, blobKey: string) {
+    const factory = this.factories.find(x => x.name === metadata.flavour);
+    if (!factory) {
+      throw new Error(`Unknown workspace flavour: ${metadata.flavour}`);
+    }
+    return factory.getWorkspaceBlob(metadata.id, blobKey);
+  }
+
+  private instantiate(metadata: WorkspaceMetadata) {
+    logger.info(`open workspace [${metadata.flavour}] ${metadata.id} `);
+    const factory = this.factories.find(x => x.name === metadata.flavour);
+    if (!factory) {
+      throw new Error(`Unknown workspace flavour: ${metadata.flavour}`);
+    }
+    const serviceCollection = this.serviceProvider.collection.clone();
```

Comment:
any advice on when to use `.clone`?

Reply:
`.clone` is quite cheap as it only clones the references.

And since the `ServiceCollection` is only about service definitions, clone it has no side effects.

Clone it here because we will inject some services specific to the current Workspace into the `ServiceCollection` in the `WorkspaceFactory`.